documentation/writing_tests.md

```diff
@@ -14,6 +20,139 @@
+- [`mocha-junit-reporter`](https://www.npmjs.com/package/mocha-junit-reporter), which produces JUnit-style XML test results.
+- [`nyc`](https://www.npmjs.com/package/nyc) is [Istanbul](https://istanbul.js.org/)'s command line interface.
+- [`esm`](https://www.npmjs.com/package/esm), a popular ECMAScript module loader.
+- [`source-map-support`](https://www.npmjs.com/package/source-map-support), which provides [source map](https://github.com/mozilla/source-map) support for stack traces in node via the [V8 stack trace API](https://github.com/v8/v8/wiki/Stack-Trace-API).
+
+A full `rush add` command that includes `mocha` and all of the previous dependencies follows:
+
+```
+rush add --dev -p mocha @types/mocha mocha-multi mocha-junit-reporter nyc esm source-map-support
+```
+
+We'll explore how we use these dependencies up next.
+
+#### Configuring Mocha
+
+Mocha is used by our test scripts in our `package.json` files. We typically invoke it with some common parameters, and some parameters that help us specify different test targets and different ways to debug the tests.
+
+If we only consider the common parameters sent to mocha, we can see the following:
+
+```
+mocha --require esm --require source-map-support/register --reporter mocha-multi --reporter-options spec=-,mocha-junit-reporter=- --timeout 180000 --full-trace
+```
+
+Let's understand what's going on:
+
+- `--require esm` requires the `esm` ECMAScript module loader.
+- `--require source-map-support/register` requires the `source-map-support/register`, to support [source map](https://github.com/mozilla/source-map) for stack traces in node.
+- `--reporter mocha-multi --reporter-options spec=-,mocha-junit-reporter=-` which allows us to both use the `spec` reporter to output a hierarchical nested view of the test cases, and also produces the JUnit-style XML test results.
+- `--timeout 180000`, which specifies the maximum time each single test case can take. More on that on the next section.
+- `--full-trace`, which enables full stack traces, since Mocha by default shortens the stack traces.
+
+That command by itself is still missing two things: the actual test files and a way to generate code coverage.
+These two missing pieces can vary depending on what we're trying to test and how we're trying to debug the tests.
+
+Code coverage can be added by placing `nyc` at the beginning of the line.
+Keep in mind that `nyc` will **obscure the stack traces**, so it's preferable to
+make separate `package.json` scripts, one for automated testing through CI, with `nyc`,
+and another one for developers running tests, without `nyc`.
+
+Then we have to point mocha to our test files. If you're **not** using `nyc`, you can point to the
+bundled test file (bundled with rollup, which we will see later), typically at `dist-test/index.node.js`.
+If you are using `nyc`, point mocha to the files built by the TypeScript compiler, normally at `dist-esm/test/*.test.js`.
```

Comment:
can you remind me why we run the test modules directly here instead of using the test bundle?

Reply:
I need to ping Harsha for this. I'll let him reply.

Reply:
nyc code coverage was broken due to the bundling, In order to get the code coverage working properly, @jeremymeng changed the integration-test command to use the dist-esm files instead.

I think @sadasant pointed that the stack trace is getting messed up with the nyc code coverage when an error is thrown in the tests making it harder to debug the sdk (or something in these lines). So, we eventually ended up having both versions of test commands.

So, we currently have two ways of testing in our package.json's,
1. integration-test : with dist-esm files, and nyc code coverage
2. unit-test : with the rollup generated bundle, no code coverage

Reply:
I guess there's no way to get nyc to import our sourcemaps?

Reply:
We tried for some time and failed. There could be a way. I wonder if @HarshaNalluru has more information.

Reply:
I haven't done a deeper investigation or looked for a workaround to get nyc work with sourcemaps, my previous comment has all the info that I could recall.